rs/https_outcalls/consensus/src/pool_manager.rs

```diff
@@ -224,6 +227,16 @@ impl CanisterHttpPoolManagerImpl {
             );
             return Vec::new();
         };
+
+        if self.active_callback_ids().len() >= MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT {
```

Comment:
Say we received `MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT` http outcalls requests at once, meaning we created that many contexts in a single round. Or more generally, say once we reach `MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT` contexts, we receive new requests faster than we can complete the existing ones. Wouldn't we never be able to get past this condition?

Reply:
Thanks Leo, great point! I planned to refactor the constant `CANISTER_HTTP_CLIENT_CHANNEL_CAPACITY` which is currently at `3000` to something lower than `MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT` because there is no way we would have this throughput anyway. But you are right, if I do it in two steps and we cut the release in the middle, there's the theoretical danger of breaking the http outcalls on a subnet.

I'll think if I can just merge them in this one PR without blowing up the diff.

Reply:
I see, but wouldn't we still have the same problem? If we ever received `MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT` or more requests contexts in the replicated state at once, we would never even try to create shares for them, regardless of the in flight channel size

Reply:
Please check the new diff:

- I simplified everything to one constant `MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT` used both for creation of new requests and for the back pressure.
- Instead of limiting the share creation, I now limit `make_new_requests` (which in turn calls `Send`), so the _shares creation_ is never restricted now.
- Note that, strictly speaking` the previous back pressure is not even really necessary, but since both the `CanisterHttpAdapterClient` and the `CanisterHttpPoolManagerImpl` have public interfaces that could lead to the increase of open calls, I let the previous restriction as is (just reduced it to 100).

Reply:
I still think if we ever find `MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT` or more request contexts in the replicated state at once, we would now never start making the http requests for them and therefore also not create any shares. So then the request contexts would just keep piling up in the replicated state without us ever starting to work on them

Reply:
> I still think if we ever find MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT or more request contexts in the replicated state at once, we would now never start making the http requests for them

But since we would create shares for the requests in the state, we would be constantly reducing the backlog, right? Remember, we only restrict creation of new http requests in this PR once we have too many, we do not restrict the processing.

> So then the request contexts would just keep piling up in the replicated state without us ever starting to work on them

I think this should not happen as the back pressure for the execution layer exists as before, because if we don't create shares, we don't empty the queue on the execution.

Reply:
Doesn't the queue go to the canister http adapter instead of execution? 
1. Execution receives new canister http request via the mgmt canister endpoint
2. Execution creates a new context for the request and adds it to the replicated state, see `http_requests` in `make_new_requests`.
3. If there are `MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT` or more contexts in the state return
4. Derive new http requests from the contexts and add them to the http adapter channel 
5. Http adapter does the http request and writes the result to the channel
6. Http client receives result from the channel and creates a share
7. Payload builder aggregates shares and puts aggregated result into block payload
8. Result is delivered back to execution via batch delivery
9. Execution removes the request context from replicated state and responds to the mgmt canister call

So now if execution adds `MAX_CANISTER_HTTP_RESPONSES_IN_FLIGHT` contexts to the state at once, we will return in 3. without ever handling the requests

Reply:
I see now that we have this middle layer `http_adapter_shim` which is used as an intermediate storage of requests by the http pool manager while talking to itself... asynchronously? This does not make sense yet and I think I need to study this code more. Thanks for your help!

Reply:
Discussed offline, decided on trying a different approach.